Feature/bokeh plots

markbotterill opened the pull request:
Reviewer: @amiraliemami 
Estimate: 1-2 hours

---

## Ticket

Fixes: [JIRA_TICKET_LINK](https://idinsight.atlassian.net/jira/software/c/projects/AAQ/boards/43/backlog?selectedIssue=AAQ-788)

## Description

Adds topic modeling visualization to the front end

### Changes made

Topic modeling pipeline now writes to Redis during running
Front end pulls out this new Json stored in the Redis and renders interactive Bokeh with it

### Future Tasks (optional)

- Improve aesthetics -> CSS for buttons, update spacing within bokeh, make sure long text from hovertip does not overflow from plot
- Add a pop-up modal that allows for the creation of on-the-fly content cards (take logic from admin app main page and put into a smaller modal)
- Add a button for a "Soft refresh" -> Requires: Save BertTopic pipeline (and eliminate all stochasticity so that it can be reloaded, new content card can be updated and a new Bokeh plot can be rendered exactly the same only with new content embedding integrated. This will enable near-live plugging of content holes that can be verified on the fly

## How has this been tested?

Runs locally + passes tests

## To-do before merge (optional)

## Checklist

Fill with `x` for completed. 

- [ x] My code follows the style guidelines of this project
- [ x] I have reviewed my own code to ensure good quality
- [ x] I have tested the functionality of my code to ensure it works as intended
- [x ] I have resolved merge conflicts

(Delete any items below that are not relevant)
- [ ] I have updated the automated tests
- [ ] I have updated the scripts in `scripts/`
- [ ] I have updated the requirements
- [ ] I have updated the README file
- [ ] I have updated affected documentation
- [ ] I have added a blogpost in Latest Updates
- [ ] I have updated the CI/CD scripts in `.github/workflows/`
- [ ] I have updated the Terraform code

amiraliemami reviewed (commented):
Leaving mid-way comments, thanks!!

amiraliemami reviewed (changes requested):
Awesome work man, looks great!

Pls add the timeout required and reorder the components as below, thanks!

<img width="824" alt="image" src="https://github.com/user-attachments/assets/65a546e4-b21f-46d5-8de3-70af028f04d6">

markbotterill commented on admin_app/src/app/dashboard/components/insights/Bokeh.tsx:
```diff
@@ -0,0 +1,219 @@
+          <Typography variant="h6">Topic Visualization</Typography>
+        </AccordionSummary>
+        <AccordionDetails>
+          {loadError && (
+            <Box sx={{ color: "error.main", mb: 2 }}>Error: {loadError}</Box>
+          )}
```
@amiraliemami I have implemented a fix for this (see commit 95d606f) - let me know if you think it's sufficient

markbotterill commented on Makefile:
```diff
@@ -39,7 +39,6 @@ guard-%:
 
 # Add users to db
 add-users-to-db:
```
Tried to do the whole branch w/o adding this as a commit 🤦 Must've slipped through - my bad.

markbotterill commented on core_backend/app/llm_call/llm_prompts.py:
```diff
@@ -410,7 +410,9 @@ class TopicModelLabellingResult(BaseModel):
         You are a summarization bot designed to condense multiple
-        messages into a topic description specific to {context}. If unknown, respond
+        messages into a topic description specific to {context}. You may
+        encounter queries that are related to something other than maternal health.
```
Used {context} again to avoid this loophole

markbotterill commented:
> Awesome work man, looks great!
> 
> Pls add the timeout required and reorder the components as below, thanks!
> 
> <img alt="image" width="824" src="https://private-user-images.githubusercontent.com/41763233/370684740-65a546e4-b21f-46d5-8de3-70af028f04d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3OTI5MjcsIm5iZiI6MTcyNzc5MjYyNywicGF0aCI6Ii80MTc2MzIzMy8zNzA2ODQ3NDAtNjVhNTQ2ZTQtYjIxZi00NmQ1LThkZTMtNzBhZjAyOGYwNGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDE0MjM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0MzU0NjJhNmMwYTQ1ZGYzOGMzNTNjMzhkNjU5OGU1MjY2YjkzODViNjQyMWQ3M2M4MTljOWVjNTU3OTIwMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wlLigQoHT3NBLyBfaX4n48tX9Y6DKfYr9cXQ6WCOE2A">

@amiraliemami I added to the Dockerfile and removed the timeout command.

suzinyou reviewed (approved):
Works well. thank you Mark! This is amazing work

suzinyou commented:
> Awesome work man, looks great!
> 
> Pls add the timeout required and reorder the components as below, thanks!
> 
> <img alt="image" width="824" src="https://private-user-images.githubusercontent.com/41763233/370684740-65a546e4-b21f-46d5-8de3-70af028f04d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4NzMyODksIm5iZiI6MTcyNzg3Mjk4OSwicGF0aCI6Ii80MTc2MzIzMy8zNzA2ODQ3NDAtNjVhNTQ2ZTQtYjIxZi00NmQ1LThkZTMtNzBhZjAyOGYwNGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAyVDEyNDMwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlMmRjYTJkNGNlMjM0ODc0YzhmNzQxNjAxODdlYzJiMmViNGE1ZDA1ZGY2NDJlMDIzZWZjYWI4MzEzMThmZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bE2OZkoQaxzfB_UWw7Qf_0_T4QjpHQB3zy7mcwqP96Y">

It looks like this now :) 

<img width="1095" alt="image" src="https://github.com/user-attachments/assets/fad9c3ae-8cfc-4cec-aa96-8e3ee7881f95">

markbotterill commented:
Thank you @suzinyou and @amiraliemami for all the feedback + help 🙏
